src/core/evaluator.js

```diff
@@ -419,28 +420,47 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
         return Promise.resolve();
       }
 
-      var nativeImageDecoderSupport = this.options.nativeImageDecoderSupport;
+      const nativeImageDecoderSupport = forceDisableNativeImageDecoder ?
+        NativeImageDecoding.NONE : this.options.nativeImageDecoderSupport;
       // If there is no imageMask, create the PDFImage and a lot
       // of image processing can be done here.
-      var objId = 'img_' + this.idFactory.createObjId();
-      operatorList.addDependency(objId);
-      args = [objId, w, h];
+      const objId = 'img_' + this.idFactory.createObjId();
 
       if (nativeImageDecoderSupport !== NativeImageDecoding.NONE &&
           !softMask && !mask && image instanceof JpegStream &&
           NativeImageDecoder.isSupported(image, this.xref, resources,
                                          this.pdfFunctionFactory)) {
         // These JPEGs don't need any more processing so we can just send it.
-        operatorList.addOp(OPS.paintJpegXObject, args);
-        this.handler.send('obj', [objId, this.pageIndex, 'JpegStream',
-                                  image.getIR(this.options.forceDataSchema)]);
-        if (cacheKey) {
-          imageCache[cacheKey] = {
-            fn: OPS.paintJpegXObject,
-            args,
-          };
-        }
-        return Promise.resolve();
+        return this.handler.sendWithPromise('obj', [
+          objId, this.pageIndex, 'JpegStream',
+          image.getIR(this.options.forceDataSchema)
+        ]).then(function() {
+          // Only add the dependency once we know that the native JPEG decoding
+          // succeeded, to ensure that rendering will always complete.
+          operatorList.addDependency(objId);
```

Comment:
I don't know this code, so correct me if I say something stupid.

Does the move of `addDependency` introduce any side effects? E.g. is it possible that by deferring the dependency, that at some point the operator list does not observe any dependencies and finishes too early?
If so, is it feasible to ensure that the dependency is resolved, even on failure?

Reply:
> Does the move of addDependency introduce any side effects?

It shouldn't since we're still adding the dependency *before* the operator here; more below...

> E.g. is it possible that by deferring the dependency, that at some point the operator list does not observe any dependencies and finishes too early?

As far as I can tell, that shouldn't be a problem since the (now async) `buildPaintImageXObject` method won't be resolved until *either* this code-path or the fallback just below has completed successfully. This should thus ensure that the `operatorList` cannot be finished before the image resources are fully parsed and handled.